Updated README.md File for Communication Module

Tejveerpratap2803 opened the pull request:
- Updated the Readme.md file for communication module.
- Added detailed sections including Overview, Architecture, System Flow Diagram, Features, Build Instructions, Project Structure, Testing, Documentation, and Contribution Guidelines to improve clarity.

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+- [Message Passing Design](score/mw/com/message_passing/design/README.md) - Low-level messaging details
+- [Safety Requirements](score/mw/com/doc/assumptions/README.md) - Safety assumptions and requirements
+
+## Contributing
+
+We welcome contributions to the Communication Module! Please read our contribution guidelines:
+
+1. **Fork** the repository
+2. **Create** a feature branch (`git checkout -b feature/amazing-feature`)
+3. **Follow** coding standards (C++17, Google style guide)
+4. **Add** tests for new functionality
+5. **Ensure** all tests pass (`bazel test //...`)
+6. **Commit** changes (`git commit -m 'Add amazing feature'`)
+7. **Push** to branch (`git push origin feature/amazing-feature`)
+8. **Open** a Pull Request
```
Have you considered putting this into a separate file as described here? https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors

I also wonder if we can remove some instructions. For example we definitely have to write that pull requests have to be from forks and that new code needs tests. But do we need to write that the contributor has to push commits to its feature branch?

And what definitely is missing is that you have to sign an eclipse contribution agreement.

Tejveerpratap2803 replied:
Thanks for the suggestions! I've moved the contribution steps to a separate CONTRIBUTING.md file, removed redundant instructions, and added the requirement to sign the Eclipse Contributor Agreement. Let me know if any other changes are needed.

lurtz commented on README.md:
````diff
@@ -1,2 +1,263 @@
+7. **Push** to branch (`git push origin feature/amazing-feature`)
+8. **Open** a Pull Request
+
+### Development Setup
+```bash
+# Setup development environment
+bazel build //...
+bazel test //...
+
+# Format code
+clang-format -i $(find . -name "*.cpp" -o -name "*.h")
+
+# Lint code
+bazel build //... --config=clang-tidy
+```
````
I was expecting this under getting started with the other bazel commands. Does it make sense to move it there?

Tejveerpratap2803 replied:
Thanks for suggestion! I will move it to reame.md file:)

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+The Communication Module (also known as **LoLa** - Low Latency) is an open-source implementation that provides:
+
+- **High-Performance Intra-ECU IPC**: Zero-copy shared-memory communication for minimal latency within ECUs
+- **AUTOSAR Compliance**: Partial implementation of Adaptive AUTOSAR Communication Management (ara::com)
```
```suggestion
```

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+- **AUTOSAR Compliance**: Partial implementation of Adaptive AUTOSAR Communication Management (ara::com)
+- **Safety-Critical**: Designed for automotive safety standards (ASIL-B qualified)
+- **Multi-Platform**: Supports Linux and QNX operating systems  
+- **Multi-Binding Architecture**: Currently implements LoLa (intra-ECU) with architectural preparation for SOME/IP (inter-ECU)
```
```suggestion
```

Tejveerpratap2803 replied:
I was thinking of adding this so that open-source contributors might get a better idea of what's currently in development. I'd love to hear your thoughts on this.

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+
+The module consists of two main components:
+
+### 1. LoLa/mw::com (High-Level Middleware)
+- **Service Discovery**: Automatic service registration and discovery
+- **Event/Field Communication**: Publish-subscribe messaging patterns  
+- **Method Invocation**: Remote procedure call (RPC) support
+- **Quality of Service**: ASIL-B and QM (Quality Management) support
+- **Zero-Copy**: Shared-memory based data exchange
+
+### 2. Message Passing (Low-Level Foundation)
+- **Asynchronous Communication**: Non-blocking message exchange
+- **Multi-Channel**: Multiple senders to single receiver communication (unidirectional n-to-1)
+- **OS Abstraction**: POSIX and QNX-specific implementations
+- **Message Types**: Support for short messages (~8 bytes payload) and medium messages (~16 bytes payload)
```
Can you add links to source code or documentation / design for each?

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+
+## Features
+
+### **Intra-ECU Communication (LoLa - Fully Implemented)**
+
+- **Zero-Copy Shared Memory**: Ultra-low latency communication within ECUs using shared memory
+- **AUTOSAR ara::com API**: Partial implementation of Adaptive AUTOSAR Communication Management
+- **Event-Driven Architecture**: Publisher/subscriber pattern with skeleton/proxy framework
+- **Service Discovery**: Flag file-based service registration and lookup mechanism
+- **Safety-Critical Design**: ASIL-B qualified implementation with deterministic behavior
+- **Multi-Threading Support**: Thread-safe operations with atomic data structures
+- **Memory Management**: Custom allocators optimized for shared memory usage
+- **Tracing Infrastructure**: Zero-copy, binding-agnostic communication tracing support
+
+### **Inter-ECU Communication (SOME/IP - Architectural Preparation)**
+
+> **Note**: Inter-ECU communication via SOME/IP is under architectural preparation phase. The features will be updated after implementation.TODO
```
this looks redundant to what was written above already

Tejveerpratap2803 replied:
Thanks for suggestion. I will make those changes:)

lurtz commented on README.md:
````diff
@@ -1,2 +1,263 @@
+bazel build //... --config=clang-tidy
+```
+
+## Project Information
+
+- **Project Home**: [Eclipse Score](https://projects.eclipse.org/projects/automotive.score)
+- **Maintainer**: Eclipse Foundation
+- **Category**: Automotive Software
+- **Language**: C++17
+- **Build System**: Bazel
+- **Platforms**: Linux, QNX
````
I would omit this at the moment, but I have no good argument.

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+## Support
+
+### Community
+- **Issues**: Report bugs and request features via GitHub Issues
+- **Discussions**: Join community discussions on the Eclipse forums
+- **Documentation**: Comprehensive docs in the `design/` and `doc/` directories
```
can you add links?

lurtz commented on README.md:
```diff
@@ -1,2 +1,263 @@
+
+[![Eclipse Score](https://img.shields.io/badge/Eclipse-Score-orange.svg)](https://eclipse-score.github.io/score/main/modules/communication/index.html)
+
+A high-performance, safety-critical communication middleware implementation based on the Adaptive AUTOSAR Communication Management specification. This module provides zero-copy, shared-memory based inter-process communication (IPC) for intra-ECU communication and for inter-ECU communication using Some/IP(future implementation) in automotive and embedded systems.
```
I just noticed that autosar is also referenced in other documents in this repo. Then it should be ok to keep mentioning it.

lurtz commented on README.md:
````diff
@@ -1,2 +1,263 @@
+│  │  • Manage connections   • Handle service lifecycle              │    │
+│  └─────────────────────────────────────────────────────────────────┘    │
+│                                    │                                    │
+│                              3. Service Match                           │
+│                                    ▼                                    │
+│  ┌─────────────────────────────────────────────────────────────────┐    │
+│  │                    Event System                                 │    │
+│  │  ┌─────────────┐           ┌─────────────┐                      │    │
+│  │  │  Publisher  │ 4. Send() │ Subscriber  │ 5. Receive()         │    │
+│  │  │   Skeleton  │◄─────────►│   Proxy     │                      │    │
+│  │  └─────────────┘           └─────────────┘                      │    │
+│  └─────────────────────────────────────────────────────────────────┘    │
+│                                    │                                    │
+│                            6. Zero-Copy Transfer                        │
+│                                    ▼                                    │
+│  ┌─────────────────────────────────────────────────────────────────┐    │
+│  │                  Shared Memory Manager                          │    │
+│  │  ┌─────────────┐  ┌─────────────┐  ┌─────────────┐              │    │
+│  │  │   Memory    │  │   Memory    │  │   Memory    │              │    │
+│  │  │   Pool 1    │  │   Pool 2    │  │   Pool N    │              │    │
+│  │  └─────────────┘  └─────────────┘  └─────────────┘              │    │
+│  └─────────────────────────────────────────────────────────────────┘    │
+└─────────────────────────────────────────────────────────────────────────┘
+                                    │                                     
+                            7. Direct Memory Access                       
+                                    ▼                                     
+┌─────────────────────────────────────────────────────────────────────────┐
+│                        Operating System                                 │
+│                      (Linux / QNX Neutrino)                             |
+└─────────────────────────────────────────────────────────────────────────┘
+
+Flow Steps:
+1. Publisher registers service with unique identifier
+2. Subscriber searches for services by identifier  
+3. Service discovery matches publisher and subscriber
+4. Publisher sends data to shared memory (zero-copy)
+5. Subscriber receives notification of new data
+6. Data transferred via direct shared memory access
+7. OS provides underlying memory management and synchronization
+```
````
But when thinking about it, I also think that it is better to describe the architecture of this repo also in this repo, but I am still new to S-CORE and do not know the reasons for doing so.

Tejveerpratap2803 replied:
Thanks for the suggestions! I created the diagram using draw.io and exported it as an SVG with embedded data for easy future edits. I felt a module-specific view in the README would help contributors, but I’m open to switching to PlantUML or linking to central docs if preferred. Appreciate the guidance!

Tejveerpratap2803 commented on README.md:
```diff
@@ -1,2 +1,263 @@
+
+## Getting Started
+
+### Prerequisites
+- **C++ Compiler**: GCC 12+ with C++17 support
+- **Build System**: Bazel 6.0+
+- **Operating System**: Linux (Ubuntu 24.04+) or QNX
+- **Dependencies**: GoogleTest, Google Benchmark
```
Thanks for the suggestions! 
I will include the devcontainer setup in the instructions and mark it as recommended so that new developers are encouraged to try it first. At the same time, I’ll keep the manual prerequisites listed for those who might not be able to use the devcontainer or prefer setting things up locally.

Tejveerpratap2803 commented on README.md:
````diff
@@ -1,2 +1,263 @@
+bazel build //score/mw/com:all
+```
+
+## Project Structure
+
+```
+communication/
+├── score/mw/com/                    # Main communication middleware
+│   ├── design/                      # Architecture documentation
+│   ├── impl/                        # Implementation details
+│   │   ├── bindings/lola/          # LoLa binding implementation
+│   │   │   ├── service_discovery/  # Service discovery client
+│   │   │   ├── messaging/          # Message passing facade
+│   │   │   └── test/               # Component tests
+│   │   └── configuration/          # Runtime configuration
+│   ├── message_passing/             # Low-level messaging
+│   │   ├── design/                 # Message passing architecture
+│   │   ├── qnx/                    # QNX-specific implementation
+│   │   └── mqueue/                 # POSIX message queue implementation
+│   ├── doc/                        # Requirements and assumptions
+│   └── example/                    # Usage examples
+├── third_party/                    # External dependencies
+├── bazel/                          # Build configuration
+└── BUILD                          # Root build file
+```
````
I understand the concern about sync issues, but I felt adding the project structure helps new developers navigate the repo better. We can mitigate the sync risk by updating the README alongside new features—open to your thoughts:)

lurtz replied:
What about this compromise?

In `score/mw/com` there is already a `README.md` and you show in this tree what is outside `score/mw/com`. In `score/mw/com/README.md` you add another tree which shows the details there.


```suggestion
## Project Structure

```text
communication/
├── score/mw/com/  # Main communication middleware (design, code, tests, examples)
├── third_party/        # External dependencies
├── bazel/                 # Build configuration
└── BUILD                # Root build file
```

and this gets added to `score/mw/com/README.md`:

```suggestion
## Project Structure
```text
├── design/                      # Architecture documentation
├── impl/                        # Implementation details
│   ├── bindings/lola/          # LoLa binding implementation
│   │   ├── service_discovery/  # Service discovery client
│   │   ├── messaging/          # Message passing facade
│   │   └── test/               # Component tests
│   └── configuration/          # Runtime configuration
├── message_passing/             # Low-level messaging
│   ├── design/                 # Message passing architecture
│   ├── qnx/                    # QNX-specific implementation
│   └── mqueue/                 # POSIX message queue implementation
├── doc/                        # Requirements and assumptions
└── example/                    # Usage examples
```

I am still afraid this will get out of sync, but now the damage is more limited

Tejveerpratap2803 commented:
- Moved contribution steps to CONTRIBUTING.md
- Added ECA requirement and build/test instructions.
- Updated README and included SVG flow diagram for intra-ECU communication.

lurtz commented on CONTRIBUTION.md:
should this file be called `CONTRIBUTING.md` instead?

Like this one: https://github.com/rails/rails/blob/main/CONTRIBUTING.md

Tejveerpratap2803 replied:
yes, I will change the name of the file:)

lurtz commented on CONTRIBUTION.md:
```diff
@@ -0,0 +1,101 @@
+# Contributing to the Communication Module
+
+Thank you for your interest in contributing to the Communication Module of the Eclipse SCORE project! This guide will walk you through the necessary steps to get started with development, adhere to our standards, and successfully submit your contributions.
```
```suggestion
Thank you for your interest in contributing to the Communication Module of the Eclipse SCORE project!
This guide will walk you through the necessary steps to get started with development, adhere to our standards, and successfully submit your contributions.
```

lurtz commented:
I do not know why, but I cannot resolve any of my findings
